src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -68,119 +74,77 @@ public boolean isRunning(){
         }
     }
     public void Stop(){
-        if(mBTConnector != null){
-            mBTConnector.Stop();
-            mBTConnector = null;
+
+        BTConnector tmpCon = mBTConnector;
+        mBTConnector = null;
+        if(tmpCon != null){
+            tmpCon.Stop();
         }
 
-        if (mServerSocketList != null) {
-            for(int i = 0; i < mServerSocketList.size(); i++) {
-                BtToServerSocket tmp = mServerSocketList.get(i);
-                if (tmp != null) {
-                    print_debug("Disconnect:::Stop : mBtToServerSocket :" + tmp.getName());
-                    tmp.Stop();
-                    tmp = null;
-                    mServerSocketList.set(i, tmp);
-                }
-            }
+        //disconnect all incoming connections
+        DisconnectIncomingConnections();
+
+        // disconnect outgoing connection
+        Disconnect("");
+    }
 
-            mServerSocketList.clear();
+    // we only cut off our outgoing connections, incoming ones are cut off from the other end.
+    // if we want to cut off whole communications, we'll do Stop
+    public boolean Disconnect(String peerId){
+        BtToRequestSocket tmpSoc = mBtToRequestSocket;
+        if(tmpSoc == null) {
+            return false;
         }
 
-        if(mBtToRequestSocket != null) {
-            print_debug("Disconnect:::Stop : mBtToRequestSocket");
-            mBtToRequestSocket.Stop();
+        String currentpeerId = tmpSoc.GetPeerId();
+        print_debug("Disconnect : " + peerId + ", current request : " + currentpeerId);
+        if (peerId.length() == 0 || peerId.equalsIgnoreCase(currentpeerId)) {
```

Comment:
The use of an empty string to mean disconnect really worries me. Mostly because this means that if someone in node.js land calls disconnect with an empty string that will also cause the current peer to disconnect. That is a great way to create really bad bugs. We should always make the programmer be explicit about their desire. So I would remove the "" flag (which anyway is ambiguous) and create a private method "DisconnectAll" that Stop can call and make Disconnect "safe" by requiring an exact peerid match.